bb8/src/internals.rs

```diff
@@ -41,12 +62,19 @@ where
         let conn = locked.conns.pop_front().map(|idle| idle.conn);
         let approvals = match &conn {
             Some(_) => locked.wanted(&self.statics),
-            None => locked.approvals(&self.statics, 1),
+            None => {
+                let approvals = min(1, locked.inflight_gets.saturating_sub(locked.pending_conns));
+                locked.approvals(&self.statics, approvals)
+            }
         };
 
         (conn, approvals)
     }
 
+    pub(crate) fn request(self: &Arc<Self>) -> GetGuard<M> {
+        GetGuard::new(self.clone())
+    }
```

Comment:
It feels like `request()` and `pop()` should be merged into a single method, which I'm thinking should yield a custom enum, maybe like this:

```rust
enum Requested {
    Connection(Conn<M::Connection>),
    Waiting(GetGuard<M>),
}
```

Reply:
This approach makes sense. It keeps tracking responsibilities more separated between inner and internals.

Reply:
Although if we put this in `pop()` we'd get the guard inside the `loop`, which means we'd lose it on the next iteration